File: Valid_BST.cpp
Time Complexity-O(n)
Space Complexity-O(1) if recursive stack space is not considered or else it will be O(h) where h=height of tree
Did the code run on Leetcode? Yes

#include <iostream>
using namespace std;

struct node  
{ 
    int data; 
    struct node *left; 
    struct node *right; 
};

struct node* newNode(int data) 
{ 
  struct node* node = (struct node*)malloc(sizeof(struct node)); 
  node->data = data; 
  node->left = NULL; 
  node->right = NULL; 
  return(node); 
} 

class Solution {
public:
    bool isValidBST(node* root) {
        return isValidUtil(root, NULL, NULL);
    }
    bool isValidUtil(node* root, node* l,node* r){
        if(root == NULL) return true;
        
        if(l != NULL && root->data >= l->data)
            return false;
        if(r != NULL && root->data <= r->data)
            return false;
        return isValidUtil(root->left,root,r) 
            && isValidUtil(root->right,l,root);
    }
};

int main() {
	struct node *root = newNode(2);
	root->left=newNode(1);
	root->right=newNode(3);
	Solution soln;
	bool result=soln.isValidBST(root);
	cout<<result;
	return 0;
}

File: ValidBST.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
// Time Complexity : O(n) number of nodes
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Your code here along with comments explaining your approach
//Recursive approachusing inorder traversal. Inorder traversal return sorted array so compare previous and currently visited val whether current is greater than previous value. make sure to return false while traversing in the left node

class Solution {
    var prev: TreeNode? = nil
    func isValidBST(_ root: TreeNode?) -> Bool {
        return inOrder(root)
    }
    
     func inOrder(_ root:TreeNode?) ->Bool{
        if root == nil {
            return true
        }
        
        if !inOrder(root?.left) {
            return false
        }
         
        if prev != nil && prev!.val >= root!.val {
            return false
        }
        prev = root
        return inOrder(root?.right) 
     }
    
  
}

// Time Complexity : O(n) Number of nodes
// Space Complexity : O(h) h is the height of the tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : missed prev = root assignment

// Your code here along with comments explaining your approach
//iterative appoach using stack inorder traversal. Inorder traversal return sorted array so compare previous and currently visited val whether current is greater than previous value. make sure to return false while traversing in the left node

class Solution {
    var prev: TreeNode? = nil
    func isValidBST(_ rootNode: TreeNode?) -> Bool {
        var prev:TreeNode? = nil
        var stack:[TreeNode] = [TreeNode]()
        var root = rootNode
        while root != nil || !stack.isEmpty {
            while root != nil {
                stack.append(root!)
                root = root?.left
            }
            root = stack.popLast()
            if prev != nil && prev!.val >= root!.val {
                return false
            }
            prev = root
            root = root?.right
        }
        return true
    }
    
}